# Tutorial on "Gene regulatory network inference from single-cell transcriptomics data" (ISMB 2022)

# Schedule of the Tutorial

*11:00 am - 6:00 pm* (Lunch Break 1:00 pm - 2:00 pm; Coffee Breaks at 3:15 - 3:30 and 4:45 - 5:00 pm)

**11:00-11:30:** Welcome, Introduction, plan for day and meet tutors/speakers

**11:30-12:15:** Talk 1 (~30min+questions)
    T. M. Murali (*Brief intro to scRNA-seq, GRN inference and BEELINE*)

**12:15-13:00:** Talk 2 (~30min+questions)
    *(tentatively AM) TENET*

**13:00-14:00:** *Lunch*

**14:00-15:15:** GRN inference method 1 (*Hands-on*)
    *PySCENIC*
    *Short introduction to scRNA-seq, methods overview &amp; principle*
    *Count matrix*
    *Jupyter notebooks with steps (~15-25min)*
    *Applying GRN inference method 1 on simulated data (workflow and let users run
themselves). Simulated data from BoolODE as well as modelled scRNA-seq
parameters*
    *Advanced users (running GRN inference on public scRNA-seq data)*

**15:15-15:30:** *Break*

**15:30-16:45:** GRN inference method 2 (Hands-on)
    *Implement TENET in BEELINE*
    *Intro to TENET and BEELINE (methods overview)*
    *Jupyter notebooks with steps (~15-25min)*
    *Inference on simulated data using BEELINE/TENET*
    *Advanced users (public scRNA-seq data; processed count matrix)*

**16:45-17:00:** *Break*

**17:00-17:30:** *Discussion:* advantages, trade-offs, considerations (scaling methods to increased features vs cells, single-cell atlases, accuracy vs sensitivity, TF-target prediction vs
edge weights etc.,)

**17:30-18:00:** Wind up, discussions, feedback and perspectives

